File: Java/题目1537：买卖股票.java

// 题目1537：买卖股票

/**
 * @author:wangzq
 * @email:wangzhenqing1008@163.com
 * @date:2015-06-30 11:01:56
 * @url:http://ac.jobdu.com/problem.php?pid=1537
 */

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.StreamTokenizer;

public class Main {
    /*
     * 1537
     */
    public static void main(String[] args) throws Exception {
        StreamTokenizer st = new StreamTokenizer(new BufferedReader(
                new InputStreamReader(System.in)));
        while (st.nextToken() != StreamTokenizer.TT_EOF) {
            int n = (int) st.nval;
            st.nextToken();
            int k = (int) st.nval;
            int array[] = new int[n + 1];
            for (int i = 1; i < n + 1; i++) {
                st.nextToken();
                array[i] = (int) st.nval;
            }
            int size = 1002;
            int dp[][] = new int[size][size];
            int result = 0;
            int tempValue = Integer.MIN_VALUE;
            for (int i = 1; i < k + 1; i++) {
                tempValue = dp[i - 1][1] - array[1];
                for (int j = 2; j < n + 1; j++) {
                    dp[i][j] = Math.max(tempValue + array[j], dp[i][j - 1]);
                    result = Math.max(dp[i][j], result);
                    tempValue = Math.max(tempValue, dp[i - 1][j] - array[j]);
                }
            }
            System.out.println(result);
        }
    }
}
/**************************************************************
 Problem: 1537
 User: wangzhenqing
 Language: Java
 Result: Accepted
 Time:600 ms
 Memory:71720 kb
 ****************************************************************/

File: C++/题目1537：买卖股票.cpp

// 题目1537：买卖股票

/**
 * @author:wangzq
 * @email:wangzhenqing1008@163.com
 * @date:2015-06-30 11:01:56
 * @url:http://ac.jobdu.com/problem.php?pid=1537
 */

import java.io.BufferedReader;
#include <stdio.h>
#include <string.h>

const int maxn = 1002;
int array[maxn];
int dp[maxn][maxn];

int max(int a, int b) {
    return a > b ? a : b;
}

int main() {
    int n, k, i;
    while (scanf("%d%d", &n, &k) != EOF) {
        for (i = 1; i < n + 1; i++) {
            scanf("%d", &array[i]);
        }
        memset(dp, 0, sizeof(dp));
        int result = 0;
        int tempValue = 99999999;
        for (i = 1; i < k + 1; i++) {
            tempValue = dp[i - 1][1] - array[1];
            for (int j = 2; j < n + 1; j++) {
                dp[i][j] = max(tempValue + array[j], dp[i][j - 1]);
                result = max(dp[i][j], result);
                tempValue = max(tempValue, dp[i - 1][j] - array[j]);
            }
        }
        printf("%d\n", result);
    }
    return 0;
}

/**************************************************************
Problem: 1537
User: wzqwsrf
Language: C++
Result: Accepted
Time:90 ms
Memory:4948 kb
****************************************************************/


                        

